src/Microsoft.OData.Core/UriParser/Resolver/ODataUriResolver.cs

```diff
@@ -313,22 +342,20 @@ public virtual IEnumerable<IEdmOperationImport> ResolveOperationImports(IEdmMode
         /// <returns>The resolved key list.</returns>
         public virtual IEnumerable<KeyValuePair<string, object>> ResolveKeys(IEdmEntityType type, IList<string> positionalValues, Func<IEdmTypeReference, string, object> convertFunc)
         {
-            var keyProperties = type.Key().ToList();
-
             // Throw an error if key size from url doesn't match that from model.
             // Other derived ODataUriResolver intended for alternative key resolution, such as the built in AlternateKeysODataUriResolver,
             // should override this ResolveKeys method.
-            if (keyProperties.Count != positionalValues.Count)
+            if (type.Key().Count() != positionalValues.Count)
             {
                 throw ExceptionUtil.CreateBadRequestError(Strings.BadRequest_KeyCountMismatch(type.FullName()));
             }
 
             var keyPairList = new List<KeyValuePair<string, object>>(positionalValues.Count);
-
-            for (int i = 0; i < keyProperties.Count; i++)
+            
+            int i = 0; 
+            foreach (IEdmProperty keyProperty in type.Key())
             {
-                string valueText = positionalValues[i];
-                IEdmProperty keyProperty = keyProperties[i];
+                string valueText = positionalValues[i++];

```

Comment:
>[i++] [](start = 51, length = 5)

random acess using '[]' in the list maybe a performance problem

Reply:
Random access is allowed in lists. It does not iterate through the entire list. LinkedList would be the one where random access would be a problem. 

---
In reply to: [362108940](https://github.com/OData/odata.net/pull/1602#discussion_r362108940) [](ancestors = 362108940)